File: java/src/tarehart/rlbot/tuning/BotLog.java
package tarehart.rlbot.tuning;

import tarehart.rlbot.time.GameTime;

import java.util.HashMap;
import java.util.Map;

public class BotLog {

    private static Map<Integer, StringBuilder> logMap = new HashMap<>();
    private static String timeStamp = "";

    public static void println(String message, int playerIndex) {

        getLog(playerIndex).append(timeStamp + message).append("\n");
        System.out.println(message);
    }

    private static StringBuilder getLog(int playerIndex) {
        if (!logMap.containsKey(playerIndex)) {
            logMap.put(playerIndex, new StringBuilder());
        }
        return logMap.get(playerIndex);
    }

    public static void setTimeStamp(GameTime time) {

        String minutes = "" + time.toMillis() / 60_000;
        String seconds = String.format("%02d", (time.toMillis() / 1000) % 60);
        timeStamp =  time.toMillis() > 0 ? "(" + minutes + ":" + seconds + ")" : "";
    }

    public static String collect(int playerIndex) {
        StringBuilder log = getLog(playerIndex);
        String contents = log.toString();
        log.setLength(0);
        return contents;
    }
}

File: src/main/java/tarehart/rlbot/tuning/BotLog.kt
package tarehart.rlbot.tuning

import tarehart.rlbot.time.GameTime
import tarehart.rlbot.ui.DisplayFlags
import java.util.*

object BotLog {

    private val logMap = HashMap<Int, StringBuilder>()
    private var timeStamp = ""

    fun println(message: String, playerIndex: Int) {

        //getLog(playerIndex).append(timeStamp + message).append("\n")
        if (DisplayFlags[DisplayFlags.BOT_LOG_IN_CONSOLE] == 1) {
            println("$timeStamp $playerIndex $message")
        }
    }

    private fun getLog(playerIndex: Int): StringBuilder {
        return logMap.getOrPut(playerIndex) { StringBuilder() }
    }

    fun setTimeStamp(time: GameTime) {
        timeStamp = "%03d".format(time.toMillis() % 1000)
    }

//    fun collect(playerIndex: Int): String {
//        val log = getLog(playerIndex)
//        val contents = log.toString()
//        log.setLength(0)
//        return contents
//    }
}
